### ISSUE TITLE
Port twisted.python.constants to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6570]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)| @wallrj reported|
|-|-|
|Trac ID|trac#6570|
|Type|task|
|Created|2013-06-13 23:00:34Z|
|Branch|https://github.com/twisted/twisted/tree/py3-constants-6570|

I tried to use constants in some new twisted.names code ([#5675](https://github.com/twisted/twisted/issues/5675)) but got an attribute error when the module is imported during the Python3 tests.

```
Traceback (most recent call last):
  File "./admin/run-python3-tests", line 30, in <module>
    unittest.main(module=None, argv=["run-python3-tests", "-v"] + testModules)
  File "/usr/lib64/python3.3/unittest/main.py", line 124, in __init__
    self.parseArgs(argv)
  File "/usr/lib64/python3.3/unittest/main.py", line 168, in parseArgs
    self.createTests()
  File "/usr/lib64/python3.3/unittest/main.py", line 175, in createTests
    self.module)
  File "/usr/lib64/python3.3/unittest/loader.py", line 137, in loadTestsFromNames
    suites = [self.loadTestsFromName(name, module) for name in names]
  File "/usr/lib64/python3.3/unittest/loader.py", line 137, in <listcomp>
    suites = [self.loadTestsFromName(name, module) for name in names]
  File "/usr/lib64/python3.3/unittest/loader.py", line 96, in loadTestsFromName
    module = __import__('.'.join(parts_copy))
  File "/home/richard/projects/Twisted/branches/edns-message-5675/twisted/names/test/test_dns.py", line 21, in <module>
    from twisted.names import dns
  File "/home/richard/projects/Twisted/branches/edns-message-5675/twisted/names/dns.py", line 1821, in <module>
    from twisted.python.constants import ValueConstant, Values
  File "/home/richard/projects/Twisted/branches/edns-message-5675/twisted/python/constants.py", line 18, in <module>
    _constantOrder = count().next
AttributeError: 'itertools.count' object has no attribute 'next'

```

Here's a little patch to fix some of the simple stuff...

```
=== modified file 'twisted/python/constants.py'
--- twisted/python/constants.py	2013-03-20 06:09:10 +0000
+++ twisted/python/constants.py	2013-06-13 22:44:26 +0000
@@ -15,7 +15,8 @@
 from operator import and_, or_, xor
 
 _unspecified = object()
-_constantOrder = count().next
+def _constantOrder(i=count()):
+    return next(i)
 
 
 class _Constant(object):
@@ -180,8 +181,8 @@
         @return: an iterator the elements of which are the L{NamedConstant}
             instances defined in the body of this L{Names} subclass.
         """
-        constants = cls._enumerants.values()
-        constants.sort(key=lambda descriptor: descriptor._index)
+        constants = sorted(cls._enumerants.values(),
+                           key=lambda descriptor: descriptor._index)
         return iter(constants)

```

But then I run into some further problems which I don't really understand - maybe related to the metaclass

```
======================================================================
ERROR: test_NULL (twisted.names.test.test_dns.EDNSMessageTestCase)
test_NULL
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/richard/projects/Twisted/branches/edns-message-5675/twisted/internet/defer.py", line 138, in maybeDeferred
    result = f(*args, **kw)
  File "/home/richard/projects/Twisted/branches/edns-message-5675/twisted/internet/utils.py", line 203, in runWithWarningsSuppressed
    reraise(exc_info[1], exc_info[2])
  File "/home/richard/projects/Twisted/branches/edns-message-5675/twisted/python/compat.py", line 286, in reraise
    raise exception.with_traceback(traceback)
  File "/home/richard/projects/Twisted/branches/edns-message-5675/twisted/internet/utils.py", line 199, in runWithWarningsSuppressed
    result = f(*a, **kw)
  File "/home/richard/projects/Twisted/branches/edns-message-5675/twisted/names/test/test_dns.py", line 2459, in test_NULL
    msg2 = dns.EDNSMessage.decode(s)
  File "/home/richard/projects/Twisted/branches/edns-message-5675/twisted/names/dns.py", line 1912, in decode
    type=MESSAGE_TYPE.lookupByValue(m.answer),
  File "/home/richard/projects/Twisted/branches/edns-message-5675/twisted/python/constants.py", line 245, in lookupByValue
    for constant in cls.iterconstants():
  File "/home/richard/projects/Twisted/branches/edns-message-5675/twisted/python/constants.py", line 184, in iterconstants
    constants = sorted(cls._enumerants.values(),
AttributeError: type object 'MESSAGE_TYPE' has no attribute '_enumerants'

======================================================================
ERROR: test_emptyQueryDecode (twisted.names.test.test_dns.EDNSMessageTestCase)
test_emptyQueryDecode
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/richard/projects/Twisted/branches/edns-message-5675/twisted/internet/defer.py", line 138, in maybeDeferred
    result = f(*args, **kw)
  File "/home/richard/projects/Twisted/branches/edns-message-5675/twisted/internet/utils.py", line 203, in runWithWarningsSuppressed
    reraise(exc_info[1], exc_info[2])
  File "/home/richard/projects/Twisted/branches/edns-message-5675/twisted/python/compat.py", line 286, in reraise
    raise exception.with_traceback(traceback)
  File "/home/richard/projects/Twisted/branches/edns-message-5675/twisted/internet/utils.py", line 199, in runWithWarningsSuppressed
    result = f(*a, **kw)
  File "/home/richard/projects/Twisted/branches/edns-message-5675/twisted/names/test/test_dns.py", line 2436, in test_emptyQueryDecode
    dns.EDNSMessage.decode(b),
  File "/home/richard/projects/Twisted/branches/edns-message-5675/twisted/names/dns.py", line 1912, in decode
    type=MESSAGE_TYPE.lookupByValue(m.answer),
  File "/home/richard/projects/Twisted/branches/edns-message-5675/twisted/python/constants.py", line 245, in lookupByValue
    for constant in cls.iterconstants():
  File "/home/richard/projects/Twisted/branches/edns-message-5675/twisted/python/constants.py", line 184, in iterconstants
    constants = sorted(cls._enumerants.values(),
AttributeError: type object 'MESSAGE_TYPE' has no attribute '_enumerants'

======================================================================
ERROR: test_repr (twisted.names.test.test_dns.EDNSMessageTestCase)
test_repr
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/richard/projects/Twisted/branches/edns-message-5675/twisted/internet/defer.py", line 138, in maybeDeferred
    result = f(*args, **kw)
  File "/home/richard/projects/Twisted/branches/edns-message-5675/twisted/internet/utils.py", line 203, in runWithWarningsSuppressed
    reraise(exc_info[1], exc_info[2])
  File "/home/richard/projects/Twisted/branches/edns-message-5675/twisted/python/compat.py", line 286, in reraise
    raise exception.with_traceback(traceback)
  File "/home/richard/projects/Twisted/branches/edns-message-5675/twisted/internet/utils.py", line 199, in runWithWarningsSuppressed
    result = f(*a, **kw)
  File "/home/richard/projects/Twisted/branches/edns-message-5675/twisted/names/test/test_dns.py", line 2285, in test_repr
    repr(dns.EDNSMessage(1)),
  File "/home/richard/projects/Twisted/branches/edns-message-5675/twisted/python/util.py", line 578, in __str__
    r.append(' %s=%r' % (attr, getattr(self, attr)))
  File "/home/richard/projects/Twisted/branches/edns-message-5675/twisted/python/constants.py", line 43, in __repr__
    return "<%s=%s>" % (self._container.__name__, self.name)
AttributeError: 'NoneType' object has no attribute '__name__'

```

Attachments:

* [py3-constants-6570-docstringsreflow.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/11b/11b4d8774664258672afd6a18c2ff6f7932d3ab7/d359abde088d6caa6bc900a3e77a903c1b2b8265.diff) (2260 bytes) - added by hawkowl on 2013-09-20 16:24:44Z - Patch for 6570 that fixes up some docstrings

<details><summary>Searchable metadata</summary>

```
trac-id__6570 6570
type__task task
reporter__rwall rwall
priority__normal normal
milestone__ 
branch__branches_py3_constants_6570 branches/py3-constants-6570
branch_author__rwall rwall
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1371164434000000 1371164434000000
changetime__1381740442000000 1381740442000000
version__None None
owner__rwall rwall

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_1"></a>|@wallrj commented|
|-|-|

(In [40016]) Branching to 'py3-constants-6570'

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_2"></a>|@wallrj commented|
|-|-|

Ready for review in log:branches/py3-constants-6570
 * I used the hack described here: http://mikewatkins.ca/2008/11/29/python-2-and-3-metaclasses/ (thanks for the link itamar)
 * But two tests are still failing on Python3:
   * http://buildbot.twistedmatrix.com/builders/python-3.3-tests/builds/1442/steps/shell/logs/stdio
 * Need some advice on how to fix those two failures.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

Seems related to the fact that `__nonzero__` is gone - http://docs.python.org/3/reference/datamodel.html#object.__bool__

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

Separately, maybe your editor's max column value is set too small.  There doesn't seem to be a reason to rewrap most of those docstrings.  Is there?

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_5"></a>|@wallrj commented|
|-|-|

Thanks exarkun,

You were right about __nonzero__. I added __bool__ as an alias (as done in t.p.loopback) and all tests now pass.

I also reverted all but one of re-wrapped docstrings.

Added some missing docstrings -- although I'm not entirely sure about my new descriptions.

Raised [#6748](https://github.com/twisted/twisted/issues/6748) because FlagConstants with specific values don't seem to be documented.

Latest Build results:
 * http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/py3-constants-6570

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_6"></a>|@thijstriemstra commented|
|-|-|

This isn't a review but I wanted to point out that the url to the blog post will be outdated and maintenance burden so seems more sensible to omit it.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)|@hawkowl set owner to @hawkowl|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_8"></a>|@hawkowl set owner to @wallrj|
|-|-|

Hi rwall, thanks for working on this.

I've looked through and can't find anything wrong with the code. I was a bit iffy about the change from iteritems() to items(), but mithrandi pointed out that unless we're dealing with a million values, the change on py2 is inconsequential at worst, so I think that's fine.

The docstrings on _constantFactory at the very end that you added are reflowed way too early. Attached is a diff to fix them, as well as a early-reflowed docstring in FlagConstant and a too-long comment that twistedchecker picked up, if you think that fixing that up while that file is being changed is fine.

All tests pass, so I think it's good to merge, once the new docstrings are fixed.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_9"></a>|@itamarst commented|
|-|-|

Please read the porting wiki page; there's a bunch of requirements that haven't been met, e.g. no `__future__` imports.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_10"></a>|@hawkowl set owner to @hawkowl|
|-|-|

...oh huh, I didn't see that - let me rereview this then.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_11"></a>|@hawkowl set owner to @wallrj|
|-|-|

Thanks for pointing that out itamar - my bad, I totally forgot about the porting page, and just looked at passing tests. Anyway...

1. Line 312 should be isinstance(names, BaseString) or similar

2. Needs the __future__ imports, as mentioned above.

3. Docstrings should be updated to use L{bytes} rather L{str}, where appropriate.

Please fix these and then resubmit. It might be better for someone with more extensive py3 experience than I to review it after :)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_12"></a>|@itamarst commented|
|-|-|

Actually, I think `str` is correct. This isn't supposed to be bytes, this is supposed to be whatever the "native" string format is, since we're mapping them to Python attributes.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_13"></a>|@hawkowl commented|
|-|-|

Ahh, yes, that would be correct, actually. I guess then you don't want to use basestring either, because you're using native strings in both cases. This py3 porting stuff is a lot more in-depth than I originally took it to be.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_14"></a>|@wallrj removed owner|
|-|-|

Ready for another review in log:branches/py3-constants-6570

build results:
 * https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/py3-constants-6570

Replying to [thijs](#note_6):
> This isn't a review but I wanted to point out that the url to the
> blog post will be outdated and maintenance burden so seems more
> sensible to omit it.

Thijs,

I don't agree with that.

The hack is pretty obscure, so it's useful to have a link to a page
that explains how it works.
If the link breaks it will be a shame, but at least it will give a
clue to someone who's tempted to change that code.


Replying to [hawkowl](#note_8):
> Hi rwall, thanks for working on this.

> I've looked through and can't find anything wrong with the code. I
> was a bit iffy about the change from iteritems() to items(), but
> mithrandi pointed out that unless we're dealing with a million
> values, the change on py2 is inconsequential at worst, so I think
> that's fine.

Yeah, I don't think it will be a problem.

The Python3 porting page also says we should wrap calls to items,
values in list to prevent changes in the API return value.  But I
couldn't find any examples of those being returned, so it wasn't
required.

> The docstrings on _constantFactory at the very end that you added
> are reflowed way too early. Attached is a diff to fix them, as well
> as a early-reflowed docstring in FlagConstant and a too-long comment
> that twistedchecker picked up, if you think that fixing that up
> while that file is being changed is fine.

My wrapping column was set wrong. I've fixed it now and reflowed only
the bits that I've touched in this branch.

Replying to [itamar](#note_9):
> Please read the porting wiki page; there's a bunch of requirements that haven't been met, e.g. no `__future__` imports.

Done.

Replying to [hawkowl](#note_13):
> Ahh, yes, that would be correct, actually. I guess then you don't
> want to use basestring either, because you're using native strings
> in both cases. This py3 porting stuff is a lot more in-depth than I
> originally took it to be.

Discussed with hawkowl on irc and decided that the use of str in
constants is fine.

-RichardW.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

> > This isn't a review but I wanted to point out that the url to the blog post will be outdated and maintenance burden so seems more sensible to omit it.

> Thijs,

> I don't agree with that.

> The hack is pretty obscure, so it's useful to have a link to a page that explains how it works. If the link breaks it will be a shame, but at least it will give a clue to someone who's tempted to change that code.

It's not "If" it's "When", sadly.  Why don't we just include the information somewhere we control so it doesn't disappear?  Or link to the Python documentation that explains this?  (I'm sure Python *must* have documentation about how metaclasses work!)

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_16"></a>|@wallrj commented|
|-|-|

Replying to [exarkun](#note_15):
> It's not "If" it's "When", sadly.  Why don't we just include the information somewhere we control so it doesn't disappear?  Or link to the Python documentation that explains this?  (I'm sure Python *must* have documentation about how metaclasses work!)

Ok. The truth is I didn't really understand what the hack was doing, so couldn't explain it.

I've spent some time this afternoon understanding metaclasses and have now modified the hack to be more in keeping with the one in the article.
ie I call type in the class definition to dynamically generate an intermediate baseclass with the correct metaclass.

That required a further change to _ConstantsContainerType.__new__ which was assuming that all its classes would have a _constantType attribute -- which the new dynamic class doesn't.

I've also attempted to explain the hack and linked to relevant official Python documentation.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_17"></a>|@wallrj commented|
|-|-|

This is currently blocking [#5446](https://github.com/twisted/twisted/issues/5446).

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_18"></a>|@wallrj commented|
|-|-|

Also blocking [#6750](https://github.com/twisted/twisted/issues/6750)

In that ticket, glyph wrote a metaclass decorator. r40289

We considered doing the same in this branch, based on six.add_metaclass, but decided that it wasn't necessary.
 * http://pythonhosted.org/six/#six.add_metaclass

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_19"></a>|@glyph commented|
|-|-|

Replying to [rwall](#note_18):
> Also blocking [#6750](https://github.com/twisted/twisted/issues/6750)
> 
> In that ticket, glyph wrote a metaclass decorator. r40289
> 
> We considered doing the same in this branch, based on six.add_metaclass, but decided that it wasn't necessary.
>  * http://pythonhosted.org/six/#six.add_metaclass

I'm definitely planning on chucking that code, or at least removing it from that branch; I wrote it to unblock myself because it was less effort than even searching for this ticket, let alone integrating this branch :-).

Looking at the alternate strategy that this branch selected for metaclass declaration though, and the lengthy comment which is used to explain what the hack is doing, I would much rather have the metaclass decorator which can have its own docstring and tests.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_20"></a>|@glyph set owner to @wallrj|
|-|-|

OK, so, I don't think the metaclass thing is ideal - I don't want to have a 5-line comment every time we use a metaclass - but we can easily address this later.  Please go ahead and land this.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_21"></a>|@wallrj set status to `assigned`|
|-|-|

Here's a record of the conversation. Looking back on it, there wasn't a clear
conclusion, but I decided to keep the patch as small as possible because I
wanted it reviewed quickly.

 $ less .znc/moddata/log/rwall_freenode_#twisted-dev_20130925.log

 * 13:16:06 <rwall> exarkun: I went looking for authoritative info about the py3 metaclass thing and found this. http://pythonhosted.org/six/#six.add_metaclass
 * 13:16:25 <rwall> Would you rather I copied that into twisted.python.compat?
 * 13:18:30 <rwall> https://bitbucket.org/gutworth/six/src/69df1f152215a67e8942f8616772923308e33b80/six.py?at=default#cl-573
 * 13:19:22 <exarkun> The implementation looks more complicated than constants.py actually needs
 * 13:19:52 <exarkun> I have trouble deciding between these two ideas:
 * 13:20:08 <exarkun> 1) the code should be written in a way that is easy for any python programmer to understand and maintain
 * 13:20:44 <exarkun> 2) metaclasses are not really that complicated and any python programmer should understand how classes are created
 * 13:21:27 * fijal opts for 1)
 * 13:27:09 <exarkun> rwall: okay.  if the license is compatible then copying add_metaclass (and its unit tests) and using it sounds fine I suppose.
 * 13:27:10 <exarkun> rwall: when I commented on the ticket, I was only thinking that two or three sentences of explanation should replace the link in the comment, though
 * 13:27:35 <dstufft> six.py is BSD IIRC
 * 13:31:53 <rwall> exarkun: Ok. I'll keep it simple and modify the comment. But I have to admit that I don't fully understand metaclasses yet.
 * 18:44:44 <kenaan> ^C03rwall ^C10r40112^C14 ^C07M(branches/py3-constants-6570/twisted/python/constants.py)^O: Use a metaclass workaround closer to the original article and the one in six. Attempt to explain the workaround. ...
 * 18:45:02 <rwall> exarkun, itamar: ^^ how about that?
 * 18:47:42 <exarkun> I guess /this/ layer of metaclasses is kind of a rathole
 * 18:51:45 <exarkun> rwall: That code is disgusting.  But probably only about as disgusting as is necessary.
 * 18:51:59 <rwall> Thanks :)
 * 18:51:59 <exarkun> rwall: I'll just suggest another name than "Py3TmpBase
 * 18:52:01 <exarkun> "
 * 18:52:20 <rwall> Oh that's it?
 * 18:52:27 <rwall> Yeah I can do that.
 * 18:53:03 <rwall> I don't really know what effect that intermedite baseclass has? Is the name seen anywhere?
 * 18:53:12 <exarkun> I think the only other option would be to assemble the class entirely manually (define a bunch of methods, construct the attribute dict, then create _ConstantContainer directly with a call to type - but I don't think we want to go down that path)
 * 18:53:16 <rwall> Also not sure what pydoctor will make of it.
 * 18:53:37 <rwall> That's what they do here: https://bitbucket.org/gutworth/six/src/69df1f152215a67e8942f8616772923308e33b80/six.py?at=default#cl-573
 * 18:53:43 <exarkun> The name can be found out, that's all.
 * 18:53:54 <exarkun> The class persists forever, so "Tmp" seems inappropriate
 * 18:54:01 <exarkun> and it gets used on Python 2, so "Py3" seems inappropriate :)
 * 18:54:02 <rwall> ok
 * 18:54:28 <exarkun> rwall: well yea sort of.  They still use the class syntax to create a class, then they rip its insides out and make a new one to replace it.
 * 18:54:57 <exarkun> imagine `def foo(self): ...; def bar(self): ...; FooBar = type('FooBar', (object,), dict(foo=foo, bar=bar))´
 * 18:55:31 <exarkun> (So I guess there are multiple other options, as there usually are.  still, they all suck)
 * 18:56:10 <rwall> Cool. That looks like composition.
 * 18:56:10 <exarkun> I think pydoctor will be unable to figure out the base class.  Which I don't think matters in this case.  No other consequences jump out at me,
 * 18:56:16 <exarkun> rwall: heh
 * 18:56:40 <rwall> ok.
 * 18:58:15 <rwall> Oh one other thing. The "getattr" change won't change the "no ConstantsContainer" subsubclasses rule will it?
 * 19:00:43 <kenaan> ^C03rwall ^C10r40113^C14 ^C07M(branches/py3-constants-6570/twisted/python/constants.py)^O: use an empty name for the intermediate base class ...
 * 19:07:42 <exarkun> rwall: is there a unit test for that?  there should be
 * 19:07:48 <exarkun> rwall: I don't think it will, but it should be tested.
 * 19:08:54 <rwall> Yeah it's ok. The _ConstantsContainer.__new__ method takes care of it.
 * 19:09:47 <rwall> and there are tests for it.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_22"></a>|@wallrj set status to `closed`|
|-|-|

(In [40299]) Merge py3-constants-6570

Author: rwall
Reviewers: exarkun, hawkowl, glyph
Fixes: [#6570](https://github.com/twisted/twisted/issues/6570)

twisted.python.constants is now compatible with Python3
